nestkernel/io_manager_impl.h

```diff
@@ -27,28 +27,35 @@
 
 namespace nest
 {
-template < class RBType >
+
+template < class RecordingBackendT >
 void
-IOManager::register_recording_backend( Name name )
+IOManager::register_recording_backend( const Name name )
 {
-  RBType* recording_backend = new RBType();
-  recording_backend->pre_run_hook();
+  if ( recording_backends_.find( name ) != recording_backends_.end() )
+  {
+    throw BackendAlreadyRegistered( name.toString() );
+  }
 
+  RecordingBackendT* recording_backend = new RecordingBackendT();
+  recording_backend->pre_run_hook();
   recording_backends_.insert( std::make_pair( name, recording_backend ) );
 }
 
-template < class SBType >
+template < class StimulationBackendT >
 void
-IOManager::register_stimulation_backend( Name name )
+IOManager::register_stimulation_backend( const Name name )
 {
-  SBType* stimulation_backend = new SBType();
-  auto it = stimulation_backends_.find( name );
-  if ( it == stimulation_backends_.end() )
+  if ( stimulation_backends_.find( name ) != stimulation_backends_.end() )
   {
-    stimulation_backend->pre_run_hook();
-    stimulation_backends_.insert( std::make_pair( name, stimulation_backend ) );
+    throw BackendAlreadyRegistered( name.toString() );
   }
+
+  StimulationBackendT* stimulation_backend = new StimulationBackendT();
+  stimulation_backend->pre_run_hook();
+  stimulation_backends_.insert( std::make_pair( name, stimulation_backend ) );
 }
-}
+
+} // namespace nest
```

Comment:
I would do the following:

```cpp
Class IOManager 
{
    template<class BackendT> std::map<Name, BackendT> get_backend(const BackendT& backend);
    template<class BackendT> void register_backend(const Name& name);

};

 template<> std::map<Name, StimulationBackend> IOManager::get_backend(const StimulationBackend& ){ return stimulation_backends_;}

 template<> std::map<Name, RecordingBackend> IOManager::get_backend(const RecordingBackend& ){ return recording_backends_;}
```
And then you have only one `register_backend` function that does the same for both recording and stimulation devices, depending on the type for the template class.

Reply:
I am also unhappy about the code duplication for the different backends, but would like to leave this for now.

My hope is that we can implement a generic `ComponentRegistry` later on, which would then support everything that can be dynamically registered (e.g. neuron and synapse models, RNG types, structural plasticity growth curves, connection rules, recording and stimulation backends).

Reply:
That's fine I think. As long as we add that to a kernel-team todo-list somewhere. We should start collecting all our ideas for the next hackathon following the SLI-purge one.

Reply:
I have captured my ideas in #2642. Please feel free to comment there in case you have further ideas.